src/raftstore/store/worker/snap.rs

```diff
@@ -49,47 +60,101 @@ quick_error! {
-        let raw_snap;
-        let key;
-
-        {
-            let storage = task.storage.rl();
-            raw_snap = storage.raw_snapshot();
-            let applied_idx = storage.applied_index();
-            let term = box_try!(storage.term(applied_idx));
-            key = SnapKey::new(storage.get_region_id(), term, applied_idx);
+        let raw_snap = Snapshot::new(self.db.clone());
+
+        if *state.rl() != SnapState::Generating {
+            // task has been canceled.
+            return Ok(());
         }
 
-        match store::do_snapshot(self.mgr.clone(), &raw_snap, key.clone()) {
-            Ok(snap) => task.storage.wl().snap_state = SnapState::Snap(snap),
+        let res = store::do_snapshot(self.mgr.clone(), &raw_snap, region_id);
+        let mut state = state.wl();
+        if *state != SnapState::Generating {
+            // task has been canceled.
+            return Ok(());
+        }
+        match res {
+            Ok(snap) => {
+                *state = SnapState::Snap(snap);
+                Ok(())
+            }
             Err(e) => {
-                return Err(Error::Other(box e));
+                *state = SnapState::Failed;
+                Err(Error::Other(box e))
             }
         }
-        Ok(())
     }
-}
 
-impl Runnable<Task> for Runner {
-    fn run(&mut self, task: Task) {
+    fn handle_gen(&self, region_id: u64, state: Arc<RwLock<SnapState>>) {
         metric_incr!("raftstore.generate_snap");
         let ts = Instant::now();
-        if let Err(e) = self.generate_snap(&task) {
+        if let Err(e) = self.generate_snap(region_id, state) {
             error!("failed to generate snap: {:?}!!!", e);
-            task.storage.wl().snap_state = SnapState::Failed;
             return;
         }
         metric_incr!("raftstore.generate_snap.success");
         metric_time!("raftstore.generate_snap.cost", ts.elapsed());
     }
+
+    fn apply_snap(&self, snap_key: SnapKey) -> Result<(), Error> {
+        info!("begin apply snap data for {}", snap_key);
+        let snap_file = box_try!(self.mgr.rl().get_snap_file(&snap_key, false));
+        self.mgr.wl().register(snap_key.clone(), SnapEntry::Applying);
+        defer!({
+            self.mgr.wl().deregister(&snap_key, &SnapEntry::Applying);
+            snap_file.delete();
+        });
+        if !snap_file.exists() {
+            return Err(box_err!("missing snap file {}", snap_file.path().display()));
+        }
+        box_try!(snap_file.validate());
+        let mut reader = box_try!(File::open(snap_file.path()));
+
+        let timer = Instant::now();
+        // Write the snapshot into the region.
+        loop {
+            // TODO: avoid too many allocation
+            let key = box_try!(reader.decode_compact_bytes());
+            if key.is_empty() {
+                break;
+            }
+            let value = box_try!(reader.decode_compact_bytes());
+            box_try!(self.db.put(&key, &value));
```

Comment:
is using a batch for multiply paris faster than using put every time?

/cc @zhangjinpeng1987